File: app/src/main/java/info/horriblesubs/sher/adapter/DownloadAdapter.java
package info.horriblesubs.sher.adapter;

import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.button.MaterialButton;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import info.horriblesubs.sher.AppMe;
import info.horriblesubs.sher.R;
import info.horriblesubs.sher.api.horrible.model.ShowRelease;

public class DownloadAdapter extends RecyclerView.Adapter<DownloadAdapter.ViewHolder> {

    private final List<ShowRelease.Download> item;
    private final OnItemClick onItemClick;

    public DownloadAdapter(OnItemClick itemClick, List<ShowRelease.Download> item) {
        this.onItemClick = itemClick;
        this.item = item;
    }

    @NotNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_item_horrible_5, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, final int position) {
        holder.button.setText(Html.fromHtml(item.get(position).source));
        holder.button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onItemClick.onItemClicked(item.get(position));
            }
        });
        if (item.get(position).link == null || item.get(position).link.isEmpty())
            holder.button.setEnabled(false);
        else if (AppMe.appMe.isDark())
            holder.button.setTextColor(holder.button.getResources().getColor(android.R.color.white));
        else
            holder.button.setTextColor(holder.button.getResources().getColor(android.R.color.black));
    }

    @Override
    public int getItemViewType(int position) {
        return super.getItemViewType(position);
    }

    @Override
    public int getItemCount() {
        return item != null ? item.size() : 0;
    }

    public interface OnItemClick {
        void onItemClicked(ShowRelease.Download link);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private final MaterialButton button;

        ViewHolder(@NonNull View itemView) {
            super(itemView);
            button = itemView.findViewById(R.id.textView);
        }
    }
}

File: app/src/main/java/info/horriblesubs/sher/ui/_extras/adapters/horrible/DownloadAdapter.kt
package info.horriblesubs.sher.ui._extras.adapters.horrible

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.button.MaterialButton
import info.horriblesubs.sher.R
import info.horriblesubs.sher.ui._extras.listeners.OnItemClickListener

class DownloadAdapter(private val listener: OnItemClickListener<String>): RecyclerView.Adapter<DownloadAdapter.Holder>() {

    private var dTitles: ArrayList<String> = arrayListOf()
    private var dUrls: ArrayList<String?> = arrayListOf()

    override fun onCreateViewHolder(group: ViewGroup, viewType: Int) = Holder(
        this@DownloadAdapter,
        LayoutInflater.from(group.context).inflate(R.layout.recycler_item_h, group, false)
    )

    override fun getItemId(position: Int): Long = position.toLong()

    override fun getItemCount() = dTitles.size

    override fun onBindViewHolder(holder: Holder, position: Int) {
        holder.onBindToViewHolder(dTitles[position], dUrls[position], position)
    }

    fun reset(titles: ArrayList<String>, urls: ArrayList<String?>) {
        dTitles.clear()
        dTitles.addAll(titles)
        dUrls.clear()
        dUrls.addAll(urls)
        notifyDataSetChanged()
    }

    class Holder (val adapter: DownloadAdapter, view: View): RecyclerView.ViewHolder(view) {

        private val button: MaterialButton = itemView.findViewById(R.id.button)

        fun onBindToViewHolder(t: String, u: String?, position: Int) {
            button.setOnClickListener{
                adapter.listener.onItemClick(it, u, position)
                Log.d("OnClick", "$t: $u")
            }
            button.isEnabled = !u.isNullOrBlank()
            button.text = t
        }
    }
}